docs/articles/getting-started/execution.md

```diff
@@ -91,19 +91,21 @@ Our next entry point introduces a more complex API to allow for more performance
 
 ### Builders
 
-You provide a input and output **builders**, functions that are called on each worker and will produce the input and handle the output for that worker. The input builder function should return an iterable that yields `(epoch, item)` tuples. The output builder function should return a callback **output handler** function that can be called with each epoch and item of output produced.
+You provide a input and output **builders**, functions that are called on each worker and will produce the input and handle the output for that worker. The input builder function should return an iterable that yields `Emit(item)` or `AdvanceTo(epoch)`. The output builder function should return a callback **output handler** function that can be called with each epoch and item of output produced.
+
+The input builder also recieves the epoch to resume processing on in the case of restarting the dataflow. The input builder must somehow skip ahead in its input data to start at that epoch.
```

Comment:
If I am using kafka and I am using a windowed approach to my epochs. Does that mean that I have to track my minimum offset associated with that epoch somewhere else so that I can restart correctly from the minimum offset associated with the most recently completed window. Or am I misunderstanding?

Reply:
This is a very good point. Perhaps. Recovery requires your input builder to be able to replay all input from a specific epoch (the one passed in in `resume_epoch`). So if you're reading from Kafka you effectively need a function like `offset_for(epoch)`. If you're windowing into fixed size buckets (10 msg / epoch), then it'd be possible to: `resume_offset = resume_epoch * 10`. But if you're windowing by time, then there's no deterministic way to go from epoch back to offset, and yes you'd have to keep track of this yourself, which is annoying.

Idea: perhaps we could have something like "annotated epochs". Where an epoch is kind of an `(int, Any)` tuple. Bytewax instructs Timely to only use the `int` part for ordering, but we persist the `Any` part as well and it's metadata that you can use for recovery.

(This also could be the system that'd enable something like transactional writes on output: you put the txn ID into the epoch metadata and the output handler can close the txn.) (I also think this would be unrelated to Timely's multi-dimensional timestamp / epoch support, even though I talk about it as a 2-tuple here.)

Thoughts on that? Let me have a think about this because this is will be a frustrating and very common problem.

Reply:
I think we'll tackle this in more detail when we look at prefab input builders.